### PR TITLE
add mattermost license

### PR DESCRIPTION
**Pull request by cpanato**
- Add a mattermost license using the secret.

### CODE REVIEW

File: pkg/controller/clusterinstallation/utils.go [hunk 1]
```diff
@@ -112,14 +112,21 @@ func (r *ReconcileClusterInstallation) checkClusterInstallation(mattermost *matt
...
+func (r *ReconcileClusterInstallation) checkSecret(secretName, keyName, namespace string) (bool, error) {
 	foundSecret := &corev1.Secret{}
 	err := r.client.Get(context.TODO(), types.NamespacedName{Name: secretName, Namespace: namespace}, foundSecret)
 	if err != nil {
-		return errors.Wrap(err, "Error getting secret")
+		return false, errors.Wrap(err, "Error getting secret")
 	}
 
-	return nil
+	for key := range foundSecret.Data {
```
Comment (gabrieljackson) [inline pkg/controller/clusterinstallation/utils.go]: This should probably be:
```suggestion
	for _, key := range foundSecret.Data {
```

**Comment by cpanato** [comment]
ready to go @gabrieljackson

File: pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go [hunk 2]
```diff
@@ -308,6 +308,33 @@ func (mattermost *ClusterInstallation) GenerateDeployment(dbUser, dbPassword str
 		},
 	}
 
+	// Mattermost License
+	volumeLicense := []corev1.Volume{}
+	volumeMountLicense := []corev1.VolumeMount{}
+	if mattermost.Spec.MattermostLicenseSecret != "" {
+
```
Comment (gabrieljackson) [inline pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go]: nit: extra line

File: pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go [hunk 3]
```diff
@@ -308,6 +308,33 @@ func (mattermost *ClusterInstallation) GenerateDeployment(dbUser, dbPassword str
...
+		})
+
+		volumeMountLicense = append(volumeMountLicense, corev1.VolumeMount{
+			MountPath: "/mattermost-license",
+			Name:      "mattermost-license",
+			ReadOnly:  true,
+		})
+
+		volumeLicense = append(volumeLicense, corev1.Volume{
+			Name: "mattermost-license",
```
Comment (gabrieljackson) [inline pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go]: Just checking: should this be called `license`?

I see our full path in the env var is `/mattermost-license/license`.

File: pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go [hunk 4]
```diff
@@ -173,7 +173,7 @@ func (mattermost *ClusterInstallation) GenerateIngress() *v1beta1.Ingress {
 }
 
 // GenerateDeployment returns the deployment spec for Mattermost
-func (mattermost *ClusterInstallation) GenerateDeployment(dbUser, dbPassword string, externalDB bool, minioService string) *appsv1.Deployment {
+func (mattermost *ClusterInstallation) GenerateDeployment(dbUser, dbPassword string, externalDB, isLicensed bool, minioService string) *appsv1.Deployment {
```
Comment (gabrieljackson) [inline pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go]: We pass in `isLicensed` here, but don't actually use the value. Should we use that instead on line 314?

File: pkg/controller/clusterinstallation/controller.go [hunk 5]
```diff
@@ -197,7 +197,8 @@ func (r *ReconcileClusterInstallation) Reconcile(request reconcile.Request) (rec
 
 func (r *ReconcileClusterInstallation) checkDatabase(mattermost *mattermostv1alpha1.ClusterInstallation, reqLogger logr.Logger) error {
 	if mattermost.Spec.DatabaseType.ExternalDatabaseSecret != "" {
-		return r.checkSecret(mattermost.Spec.DatabaseType.ExternalDatabaseSecret, mattermost.Namespace)
+		_, errSecret := r.checkSecret(mattermost.Spec.DatabaseType.ExternalDatabaseSecret, "externalDB", mattermost.Namespace)
```
Comment (gabrieljackson) [inline pkg/controller/clusterinstallation/controller.go]: nit:
```suggestion
		_, err := r.checkSecret(mattermost.Spec.DatabaseType.ExternalDatabaseSecret, "externalDB", mattermost.Namespace)
```

File: pkg/controller/clusterinstallation/controller.go [hunk 6]
```diff
@@ -197,7 +197,8 @@ func (r *ReconcileClusterInstallation) Reconcile(request reconcile.Request) (rec
...
+		return errSecret
```
Comment (gabrieljackson) [inline pkg/controller/clusterinstallation/controller.go]: nit:
```suggestion
		return err
```

File: pkg/controller/clusterinstallation/utils.go [hunk 7]
```diff
@@ -112,14 +112,21 @@ func (r *ReconcileClusterInstallation) checkClusterInstallation(mattermost *matt
...
+		if keyName == key {
+			return true, nil
+		}
+	}
+
+	msg := fmt.Sprintf("Missing required secret data. Want: %s", keyName)
+	return false, errors.Wrap(err, msg)
```
Comment (gabrieljackson) [inline pkg/controller/clusterinstallation/utils.go]: If you are choosing to return an error here if the value isn't found, then there isn't much point to also returning a bool as well.

Couldn't we just infer that if there is no error, then the secret key was found?

File: pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go (hunk 4 above)
Comment (cpanato) [reply pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go]: my brain was melted yesterday, thanks for the catch. replaced

File: pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go (hunk 3 above)
Comment (cpanato) [reply pkg/apis/mattermost/v1alpha1/clusterinstallation_utils.go]: this is just the name to match with the volume. but yes the path will be `/mattermost-license` to make it clear and the file is called `license`

File: pkg/controller/clusterinstallation/utils.go (hunk 7 above)
Comment (cpanato) [reply pkg/controller/clusterinstallation/utils.go]: makes sense, done

File: pkg/controller/clusterinstallation/controller.go [hunk 8]
```diff
@@ -197,7 +197,8 @@ func (r *ReconcileClusterInstallation) Reconcile(request reconcile.Request) (rec
 
 func (r *ReconcileClusterInstallation) checkDatabase(mattermost *mattermostv1alpha1.ClusterInstallation, reqLogger logr.Logger) error {
 	if mattermost.Spec.DatabaseType.ExternalDatabaseSecret != "" {
-		return r.checkSecret(mattermost.Spec.DatabaseType.ExternalDatabaseSecret, mattermost.Namespace)
+		err := r.checkSecret(mattermost.Spec.DatabaseType.ExternalDatabaseSecret, "externalDB", mattermost.Namespace)
+		return err
```
Comment (stylianosrigas) [inline pkg/controller/clusterinstallation/controller.go]:
```suggestion
		if err != nil {
			return errors.Wrap(err, "Error getting the external database secret.")				
		}
```

Comment (cpanato) [reply pkg/controller/clusterinstallation/controller.go]: thanks for the catch! fixed

**Comment by cpanato** [comment]
@gabrieljackson PTAL

**Pull request merged**